Better describing custom field options (#1408)
strapi/documentation: docs/developer-docs/latest/development/custom-fields.md

### Options

Custom fields can have their own settings in the Content-type builder. In
`app.customFields.register()` we can pass an additional `options` object  with the following parameters enable their use.

| Options parameter | Description                                                                     | Type                    |
| -------------- | ------------------------------------------------------------------------------- | ----------------------- |
| `base`         | Settings available in the _Base settings_ tab of the field in the Content-type Builder       | `Object` or  `Array of Objects` |
| `advanced`     | Settings available in the _Advanced settings_ tab of the field in the Content-type Builder   | `Object` or  `Array of Objects` |
| `validator`    | Validator function returning an object, used to sanitize input. Uses a [`yup` schema object](https://github.com/jquense/yup/tree/pre-v1).  | `Function`              |

> By default, Strapi has options available for it's [field model](/developer-docs/latest/backend-customization/models.md#validations) you can use them in your custom field by copying the object value from [this file](https://github.com/strapi/strapi/blob/e033275bd0915049fa4fa823f00c17ad53a8fc44/packages/core/content-type-builder/admin/src/components/FormModal/attributes/attributeOptions.js#L1) and respecting the structure below. 

Reviewer: ```suggestion
By default, Strapi has options available for its [field model](/developer-docs/latest/backend-customization/models#validations). Use them in your custom field by copying the object value from [the `attributeOptions` file](https://github.com/strapi/strapi/blob/e033275bd0915049fa4fa823f00c17ad53a8fc44/packages/core/content-type-builder/admin/src/components/FormModal/attributes/attributeOptions.js#L1) and respecting the following structure: 
```

```javascript
{
	// The option name
  name: 'required', 
	// The input type for this option
  type: 'checkbox',
	// The default value for this option
	defaultValue: true,
	// The translation object for this option's label
  intlLabel: {
    id: getTrad('form.attribute.settings.default'),
    defaultMessage: 'Default value',
  },
	// The translation object for this option's description
	intlDescription: {
    id: getTrad('form.attribute.settings.default.description'),
    defaultMessage: 'Default value description',
	}
}
```

> Should you not want to use the defaults, `options` enables the creation of Custom settings when we set up or edit a custom field in the Content-types Builder. Both `base` and `advanced` settings accept an object or an array of objects, each object being a settings section. Each settings section could include:

Reviewer: ```suggestion
If defaults are not appropriate for your custom field use case, `options` enables the creation of custom settings when setting up or editing a custom field in the Content-types Builder. Both `base` and `advanced` settings accept an object or an array of objects, each object being a settings section. Each settings section could include:
```

Reviewer 2: Also do y'all think we should mention here that base and advanced refer to the basic and advanced tabs in the CTB modal form?

Reviewer 2: > Both `base` and `advanced` settings accept an object or an array of objects

I believe they only accept an array of objects 🤔

- a `sectionTitle` to declare the title of the section as an [`IntlObject`](https://formatjs.io/docs/react-intl/)

> - and a list of `items` as an array of objects.

Reviewer 2: I think we need to make an important distinction here because you don't have to provide the sectionTitle if you want everything to be in one section.

We could call what's described in the table below a `settingItem`, base and advanced tabs can have an array of `settingItem`s and they would appear in the top of the form with no section title. 

If you want to organize your settings into sections you would then need an object with a `sectionTitle` key and an `items` key. The `items` key would take an array of `settingItem`s for that section. If you want a new section but no title (ie you just want some settings grouped together). You can set `sectionTitle` to null.

The color-picker is a good example of both of these cases: https://github.com/strapi/strapi/blob/3a33df8e7e53547d82d4fb65e9fa8395a01b8b7b/packages/plugins/color-picker/admin/src/index.js#L31

| `intlLabel`     | Translation for the label of the input                             | [`IntlObject`](https://formatjs.io/docs/react-intl/) |
| `type`          | Type of the input (e.g., `select`, `checkbox`)                     | `String`                                             |

> ::: details "type" accepts the following inputs:

Reviewer: We use the details block for code examples on this page, so for consistency, I wouldn't use it here. Since we're describing the different values accepted for `type`, I'd suggest removing this block and expanding the table above with a list.

You can use something like the following for line 224 :)
```
| `type` | Type of the input.<br />Accepts the following values:  <ul><li>'text',</li><li>'checkbox',</li><li>'checkbox-with-number-field',</li><li>'boolean-radio-group',</li><li>'select-number',</li><li>'radio-group',</li><li>'select',</li><li>'select-default-boolean',</li></li>'select-date',</li><li>'textarea-enum',</li></ul>	 | `String`
```

```javascript
[
  'text',
  'checkbox',
  'checkbox-with-number-field',
  'boolean-radio-group',
  'select-number',
  'radio-group',
  'select',
  'select-default-boolean',
  'select-date',
  'textarea-enum',	

> ]

Reviewer 2: We should probably test all of these and commit them as examples on a file in the examples/getstarted, similar to how we do for types:

https://github.com/strapi/strapi/blob/3a33df8e7e53547d82d4fb65e9fa8395a01b8b7b/examples/getstarted/src/plugins/myplugin/admin/src/index.js#L71

That could be something I do though if you don't have the time.

```
::: 

> You can also set a `defaultValue` for options , below I use the select-number component and provide a `defaultValue` of `3`.

Reviewer: I can't find "select-number" in the code example 🤔 What do you mean here?

Reviewer 2: Something like this but maybe phrased better...

```suggestion
A setting item also accepts a `defaultValue`. This allows you to initialize the value of the input. For example below I use the select-number component input and provide a `defaultValue` of `3`. When accessing the input in the CTB I should see the input pre-filled with `3`
```

```javascript
{
	// The option name
  name: 'options.grade', 
	// The input type for this option
  type: 'number',
	// The default value for this option
	defaultValue: 3,
	// The translation object for this option's label
  intlLabel: {
    id: getTrad('form.attribute.settings.default'),
    defaultMessage: 'Default value',
  },
	// The translation object for this option's description
	intlDescription: {
    id: getTrad('form.attribute.settings.default.description'),
    defaultMessage: 'Default value description',

> 	}

Reviewer 2: I think the label and description would still be grade
